crypto/pem/pem_info.c

```diff
@@ -292,3 +293,84 @@ STACK_OF(X509_INFO) *PEM_X509_INFO_read_bio(BIO *bp, STACK_OF(X509_INFO) *sk,
   OPENSSL_free(data);
   return ret;
 }
+
+// A TJH addition
+int PEM_X509_INFO_write_bio(BIO *bp, X509_INFO *xi, EVP_CIPHER *enc,
+                            unsigned char *kstr, int klen, pem_password_cb *cb,
+                            void *u) {
+  int i, ret = 0;
+  unsigned char *data = NULL;
+  const char *objstr = NULL;
+  char buf[PEM_BUFSIZE];
+  unsigned char *iv = NULL;
+  unsigned iv_len = 0;
+
+  if (enc != NULL) {
+    iv_len = EVP_CIPHER_iv_length(enc);
+    objstr = OBJ_nid2sn(EVP_CIPHER_nid(enc));
+    if (objstr == NULL) {
+      OPENSSL_PUT_ERROR(PEM, PEM_R_UNSUPPORTED_CIPHER);
+      goto err;
+    }
+  }
+
+  // now for the fun part ... if we have a private key then we have to be
+  // able to handle a not-yet-decrypted key being written out correctly ...
+  // if it is decrypted or it is non-encrypted then we use the base code
+  if (xi->x_pkey != NULL) {
+    if ((xi->enc_data != NULL) && (xi->enc_len > 0)) {
+      if (enc == NULL) {
+        OPENSSL_PUT_ERROR(PEM, PEM_R_CIPHER_IS_NULL);
+        goto err;
+      }
+
+      // copy from weirdo names into more normal things
+      iv = xi->enc_cipher.iv;
+      data = (unsigned char *)xi->enc_data;
+      i = xi->enc_len;
+
+      // we take the encryption data from the internal stuff rather
+      // than what the user has passed us ... as we have to match
+      // exactly for some strange reason
+      objstr = OBJ_nid2sn(EVP_CIPHER_nid(xi->enc_cipher.cipher));
+      if (objstr == NULL) {
+        OPENSSL_PUT_ERROR(PEM, PEM_R_UNSUPPORTED_CIPHER);
+        goto err;
+      }
+
+      // create the right magic header stuff
+      assert(strlen(objstr) + 23 + 2 * iv_len + 13 <= sizeof buf);
+      buf[0] = '\0';
+      PEM_proc_type(buf, PEM_TYPE_ENCRYPTED);
+      PEM_dek_info(buf, objstr, iv_len, (char *)iv);
+
+      // use the normal code to write things out
+      i = PEM_write_bio(bp, PEM_STRING_RSA, buf, data, i);
+      if (i <= 0) {
+        goto err;
+      }
+    } else {
+      // Add DSA/DH
+      // normal optionally encrypted stuff
+      if (PEM_write_bio_RSAPrivateKey(bp, xi->x_pkey->dec_pkey->pkey.rsa, enc,
```

Comment:
* Should we use the proper function to obtain the RSA key?
```
RSA* rsa = EVP_PKEY_get0_RSA(xi->x_pkey->dec_pkey);
```
* Should we check whether `xi->x_pkey->dec_pkey` or `rsa` is NULL?